### PR TITLE
ProBlym-Löser: Subensemble package

### PR DESCRIPTION
**Pull request by MauriceFromm**
## Issue

[add common.py with subensemble_generator](https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=5%3A+ProBlym-Löser&pane=issue&itemId=140580951&issue=Cortys%7CSEP-Probly-WS2526%7C217)
[Add Torch Support for SubEnsemble](https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=5%3A+ProBlym-Löser&pane=issue&itemId=140580662&issue=Cortys%7CSEP-Probly-WS2526%7C216)
[Add Flax Support for SubEnsemble](https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=5%3A+ProBlym-Löser&pane=issue&itemId=140580555&issue=Cortys%7CSEP-Probly-WS2526%7C215)
[Add test for SubEnsemble common.py](https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=5%3A+ProBlym-Löser&pane=issue&itemId=140581625&issue=Cortys%7CSEP-Probly-WS2526%7C222)
[Add test for SubEnsemble Torch](https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=5%3A+ProBlym-Löser&pane=issue&itemId=140581551&issue=Cortys%7CSEP-Probly-WS2526%7C220)
[Add test for SubEnsemble Flax](https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=5%3A+ProBlym-Löser&pane=issue&itemId=140581527&issue=Cortys%7CSEP-Probly-WS2526%7C219)
[Tutorial Notebook for SubEnsemble](https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=5%3A+ProBlym-Löser&pane=issue&itemId=141553179&issue=Cortys%7CSEP-Probly-WS2526%7C266)

## Motivation and Context

Full ensembles can significantly improve model accuracy and are an important tool for uncertainty quantification, but they are often computationally expensive — requiring multiple copies of the entire model to be trained and stored.

Subensembles offer a promising middle ground: by sharing a backbone and only duplicating the last layers, they substantially reduce parameter and memory overhead compared to full ensembles while still improving accuracy and enabling uncertainty estimates.

Because the final classification/regression layers map learned features to outputs, duplicating those heads often yields a large portion of the ensemble diversity at much lower cost.

---

## Public API Changes

-   [ ] No Public API changes
-   [x] Yes, Public API changes (Details below)

`probly.transformation.subensemble` with init, common, flax and torch

- To import the package directly: `import probly.transformation.subensemble`
- The lazy dispatch entry-point `subensemble_generator` is defined in common

Supported backend registrations:
- torch: generator registered for torch.nn.Module (`generate_torch_subensemble(base, num_heads, head, reset_params, head_layer)`)
- flax: generator registered for flax.nnx.Module (`generate_flax_subensemble(base, num_heads, head, reset_params, head_layer)`)

`probly.transformation.__init__` required **addition** of subensemble package, non-breaking.

Tests and an example notebook were added (no API change)

#### subensemble args:

- `base`: model that should be used as backbone or to create backbone and heads.
- `num_heads`: number of how many heads should be created, needs to be a positiv integer.
- `head:` optional model that is used as head and cloned num_heads times.
- `reset_params`: whether to reset the parameters of each head, default True.
- `head_layer`: number of layers used to create the head if no head model is provided.

With these args, subensemble can be called in 3 different ways:

- `subensemble(base, num_heads)`  
splits base in backbone and heads. Heads are copied num_heads times. Only the last layer is used for the heads.

- `subensemble(base, num_heads, head_layer)`
splits base in backbone and heads. Last head_layer layers are used for the heads and copies num_heads times.

- `subensemble(base, num_heads, head)`
uses base as shared backbone and copies head num_heads times.

All of these calls support reset_params.

---

## How Has This Been Tested?

Many tests were added (`tests/probly/transformation/subensemble/`) with `test_common`, `test_flax` and `test_torch`.

These were used to test the expected behaviour.

---

## Checklist

-   [x] The changes have been tested locally.
-   [ ] Documentation has been updated (if the public API or usage changes).
-   [ ] A entry has been added to [`CHANGELOG.md`](https://github.com/pwhofman/probly/blob/main/CHANGELOG.md) (if relevant for users).
-   [x] The code follows the project's [style guidelines](https://github.com/pwhofman/probly/blob/main/.github/CONTRIBUTING.md) and passes minimal code style checks.
-   [x] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

File: src/probly/transformation/subensemble/torch.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    - using an obj as shared backbone and copying the head model num_heads times.
+    Resets the parameters of each head.
+    """
+    # no head
+    if head is None:
+        head = nn.Sequential(*list(obj.children())[-head_layer:])
+        obj = nn.Sequential(*list(obj.children())[:-head_layer])
+
+    # obj and head
+    heads = ensemble(head, num_members=num_heads, reset_params=reset_params)
```
Comment (pwhofman) [inline src/probly/transformation/subensemble/torch.py]: I suspect this call is in response to our discussion regarding the connection between ensembles and subensembles. What I meant is the following: the subensemble can be seen as a generalization of the ensemble, where the subensemble is generated without base (only heads) to generate an ensemble, see e.g. https://github.com/pwhofman/probly/blob/8e6d40deaa52b4176620c8cf39661ec37de3734a/src/probly/representation/ensemble.py. You can ignore this for now and we can discuss this again at a later stage.
If this was not in response to our discussion: ignore this altogether ;).

File: tests/probly/transformation/subensemble/test_torch.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
+"""Test for torch subensemble models."""
```
Comment (pwhofman) [inline tests/probly/transformation/subensemble/test_torch.py]: I will check these tests when the changes to the subensemble are made.

File: src/probly/transformation/subensemble/torch.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    return nn.ModuleList([obj, heads])
```
[the remark this answers is not in the document]
Comment (pwhofman) [reply src/probly/transformation/subensemble/torch.py]: Yes, subensemble should return a ModuleList such that every element of this list is an object that one can pass data through. Either of your proposed implementations should allow this. In this case, I would go for the second option, as the added structure provides an easy way for the user to access the backbone or the heads later on.
